fiftyone/utils/youtubedl.py

```diff
@@ -0,0 +1,343 @@
+            and end times in seconds for downloading segments of videos
+        max_videos (None): the maximum number of videos to download from the
+            given ``urls``. By default, all ``urls`` will be downloaded
+        num_workers (None): the number of processes to use when downloading
+            individual video. By default, ``multiprocessing.cpu_count()`` is
+            used
+        ext (None): the extension to use to store the downloaded videos or a
+            list of extensions corresponding to the given ``urls``. By default,
+            the default extension of each video is used 
+        buffer_seconds (5): the number of additional seconds before and after
+            each clip to download to ensure nearest keyframes are captured if
+            `clip_segments` is provided
+
+    Returns:
+        urls of successfully downloaded videos
+        a dict of download errors and the corresponding video urls
+    """
+    num_workers = _parse_num_workers(num_workers)
+    tasks = _build_tasks_list(
+        urls, clip_segments, download_dir, ext, buffer_seconds=buffer_seconds
+    )
+
+    if max_videos is None:
+        max_videos = len(tasks)
+
+    if num_workers == 1:
+        downloaded, errors = _single_thread_download(tasks, max_videos)
+
+    else:
+        downloaded, errors = _multi_thread_download(
+            tasks, max_videos, num_workers
+        )
+
+    return downloaded, errors
+
+
+def _parse_num_workers(num_workers):
+    if num_workers is None:
+        if os.name == "nt":
+            # Default to 1 worker for Windows OS
+            return 1
+        return multiprocessing.cpu_count()
+
+    if not isinstance(num_workers, int) or num_workers < 1:
+        raise ValueError(
+            "The `num_workers` argument must be a positive integer or `None` "
+            "found %s" % str(type(num_workers))
+        )
+    return num_workers
+
+
+def _build_tasks_list(
+    urls, clip_segments, download_dir, ext, buffer_seconds=5
+):
+    if isinstance(urls, str):
+        urls = [urls]
+    if isinstance(clip_segments, tuple):
+        clip_segments = [clip_segments]
+
+    if isinstance(urls, list):
+        if download_dir is None:
+            raise ValueError(
+                "When `urls` is a list, `download_dir` is required but was found to be `None`."
+            )
+        urls = {url: None for url in urls}
+
+        etau.ensure_dir(download_dir)
+
+    num_videos = len(urls)
+
+    if clip_segments is None:
+        clip_segments = [None] * num_videos
+
+    if len(clip_segments) != num_videos:
+        raise ValueError(
+            "Found %d `clip_segments` and %d `urls`, but the lengths of these "
+            "iterables must match" % (len(clip_segments), num_videos)
+        )
+
+    ext_list = _parse_list_arg(ext, num_videos)
+    urls_list = _parse_list_arg(list(urls.keys()), num_videos)
+    paths_list = _parse_list_arg(list(urls.values()), num_videos)
```

Comment:
It would be cleaner to render your `$(title)$(ext)` patterns here so that the tasks are just `(url, path, [segment])`.

Reply:
Currently the task creation logic is partly here but partly in the download method, requiring the reader to fully understand both methods before they can make sense of what either method is doing